packages/dht/src/connection/ConnectionManager.ts

```diff
@@ -0,0 +1,379 @@
+
+                    resolve(connectivityResponseMessage)
+                })
+
+                // send connectivity request
+                const connectivityRequestMessage: ConnectivityRequestMessage = { port: this.config.webSocketPort! }
+                const msg: Message = {
+                    appId: this.CONNECTION_MANAGER_APP_ID,
+                    messageType: MessageType.CONNECTIVITY_REQUEST, messageId: 'xyz',
+                    body: ConnectivityRequestMessage.toBinary(connectivityRequestMessage)
+                }
+
+                outgoingConnection.once(ConnectionEvents.ERROR, () => {
+                    logger.trace('clientsocket error')
+                })
+
+                logger.trace('trying to send connectivity request')
+                outgoingConnection.send(Message.toBinary(msg))
+                logger.trace('connectivity request sent: ' + JSON.stringify(Message.toJson(msg)))
+            }
+        })
+    }
+
+    async start(): Promise<ConnectivityResponseMessage> {
+        if (this.started || this.stopped) {
+            throw new Err.CouldNotStart(`Cannot start already ${this.started ? 'started' : 'stopped'} module`)
+        }
+        this.started = true
+        logger.info(`Starting ConnectionManager...`)
+        // Set up and start websocket server
+        if (this.webSocketServer) {
+            this.webSocketServer.bindListeners(
+                this.handleIncomingConnectivityRequest.bind(this),
+                this.onIncomingMessage.bind(this)
+            )
+
+            await this.webSocketServer.start({ host: this.config.webSocketHost, port: this.config.webSocketPort })
+
+            return new Promise(async (resolve, reject) => {
+                // Open websocket connection to one of the entrypoints and send a CONNECTIVITY_REQUEST message
+
+                if (this.config.entryPoints && this.config.entryPoints.length > 0) {
+                    this.sendConnectivityRequest().then((response) => resolve(response)).catch((err) => reject(err))
+                }
+
+                else {
+                    // return connectivity info given in config to be used for id generation
+
+                    const connectivityResponseMessage: ConnectivityResponseMessage = {
+                        openInternet: true,
+                        ip: this.config.webSocketHost!,
+                        natType: 'open_internet',
+                        websocket: { ip: this.config.webSocketHost!, port: this.config.webSocketPort! }
+                    }
+                    resolve(connectivityResponseMessage)
+                }
+            })
+        }
+        const connectivityResponseMessage: ConnectivityResponseMessage = {
+            openInternet: false,
+            ip: 'localhost',
+            natType: 'unknown'
+        }
+        return connectivityResponseMessage
+    }
+
+    enableConnectivity(ownPeerDescriptor: PeerDescriptor): void {
+        if (!this.started || this.stopped) {
```

Comment:
We want the peerDescriptor to be set by a higher layer after the manager has started, it could be possible to make it optional to perform in automatically or pass some function that handles it to `start()`

Reply:
We could have a callback in `start` which provides the `peerDescriptor`. In that case  the private `enableConnectivity` method can be called automatically by the `start` method.

Reply:
Callback has been added to start() to move from 3-phase init to standard 2-phase one